### DISCUSSION TITLE
Multivariate (multi-channel) dataset

### DISCUSSION THREAD

**Discussion by ghost** [archinetai/audio-diffusion-pytorch#16] [General]
Hello, thank you for the great repo for audio diffusion! 
I just wanted to ask if you had any experience with multi-channel audio diffusion? 
This might be out of the scope for this repo, but I am currently trying to train a diffusion model on a 12-lead electrocardiogram sampled in 500 Hz for 10 seconds ([12, 5000] shape). However, I am having difficulty training the model, so was wondering if you might have any insight to what parameters I should attempt from your intuition or experience. 
Thank you!

**Comment by flavioschneider** [comment]
You cannot pass a length that is not a power of 2, I'd suggest you to pad the cardiograms with zeros to a length of 8192. Then since the sequence is very short you can use a smaller model (~70M params), something like this:

```py
from audio_diffusion_pytorch import AudioDiffusionModel

model = AudioDiffusionModel(
    in_channels=12,
    patch_size=4,
    kernel_sizes_init=[1, 3, 7],
    multipliers=[1, 2, 4, 4, 4],
    factors=[4, 2, 2, 2],
    num_blocks=[2, 2, 2, 2],
    attentions=[False, True, True, True],
)

# Train model with cardiograms sources
x = torch.randn(1, 12, 8192)
loss = model(x)
loss.backward() # Do this many times

# Sample 2 cardiograms given start noise
noise = torch.randn(2, 12, 8192)
sampled = model.sample(
    noise=noise,
    num_steps=10 # Suggested range: 2-50
) # [2, 12, 8192]
```

**Reply by ghost** [reply]
Hello and thank you for your fast reply!
I am training with the setting you mentioned but unfortunately my loss fluctuates between 1~2 and the sampled ecg is as follows:
![media_images_Original ECG_1223_e85244242e5c9fdae11b](https://user-images.githubusercontent.com/95455900/191266726-fe1ad818-21e4-4c19-877d-1ee7f7dc4a76.png)

![media_images_Generated ECG_1224_f0cad0bb53ddb19ba0d8](https://user-images.githubusercontent.com/95455900/191266768-eccda8bd-dea0-4bdd-a09e-e498d8ab2564.png)

The first is a random sample of original ecg and the second is sampled ecg with num_steps=50 in 10th epoch (batch size =256) 
I can train for more epochs, but I don't think it will improve the model. 
Should I try to use a larger model? Thank you

**Reply by ghost** [reply]
Actually I noticed that my dataset might not have been in the correct range for training. 
But I have a question about the "start noise", theoretically if the model is trained well we should be able to use noise = torch.randn(2, 12, 8192) for the start noise when we sample, right?

**Reply by flavioschneider** [reply]
> Should I try to use a larger model? Thank you

I don't think it's a model size problem, since the output looks like pure noise. Make sure that the sources are in the -1,1 range, and that the `diffusion_sigma_data` matches the standard deviation of your dataset - with music I found 0.1 to work well, but for egc it might be different. Let me know how the experiments go, also interested in seeing if the model adapts well to other domains (in theory it should) 

>  if the model is trained well we should be able to use noise = torch.randn(2, 12, 8192) for the start noise when we sample, right?

That's correct, the start noise should be normally distributed independently of other parameters

**Reply by ghost** [reply]
Hello, and thanks for the replies!
Currently I am still trying to make this work, but the generated ECGs do not reflect the ground-truth ECGs 
Below is an example of a generated ECG  
![media_images_Generated ECG_898_cbee37f27bdbb872659a](https://user-images.githubusercontent.com/95455900/191802754-baaa9593-a58c-4ce2-b95f-f9ef1729c560.png)

Do you have any idea what might be the problem? I made sure the sources are in the -1 1 range and the diffusion_sigma_data

**Reply by flavioschneider** [reply]
Is that plot in scale? If so, something is way off. The network values are clipped to be in the range [-1,1] so that plot should not be a possible network output. Might want to check that your data, other than that it's hard to tell without more information.

**Reply by ghost** [reply]
> Is that plot in scale? If so, something is way off. The network values are clipped to be in the range [-1,1] so that plot should not be a possible network output. Might want to check that your data, other than that it's hard to tell without more information.

Sorry for the confusion, the generated data was first normalized in the -1 to 1 range with minmaxscaler and after training, the samples were displayed with inverse_transform back to its raw waveform. Like you mentioned, i think something definitely went wrong during this process. 

![media_images_Generated ECG_14918_d2c62eddb9df3a440a22](https://user-images.githubusercontent.com/95455900/191901255-5339a8c0-78a6-4054-9b7a-7104724d02e2.png)

Actually I am getting much better generated samples from training the raw waveform, where most are in the -1 to 1 range anyway excluding some outliers. Would it make sense if I sample without clipping to the [-1,1] range?

And im also wondering if the conditional generation (text-conditioned) is ready for use!

 Thanks again for your help.

**Reply by flavioschneider** [reply]
What I would do in this case is find out what are the theoretical min `p` and max `q` possible value for a human ecg, remove all outliers that execeed that (if there are any) and normalize all ecg with a fixed scale e.g. `s=max(abs(p), q)`, `ecg=ecg/s`. Then after sampling, undo the scaling with `ecg=ecg*s` – assuming those are zero centered. 

Conditional generation is ready, I haven't tested it much though. Check out the `AudioDiffusionConditional` in the readme. You have to provide an embedding, ideally not directly text but some learned embedding from a transformer output.

**Reply by ghost** [reply]
> What I would do in this case is find out what are the theoretical min `p` and max `q` possible value for a human ecg, remove all outliers that execeed that (if there are any) and normalize all ecg with a fixed scale e.g. `s=max(abs(p), q)`, `ecg=ecg/s`. Then after sampling, undo the scaling with `ecg=ecg*s` – assuming those are zero centered.

![generated_diffusion_normal](https://user-images.githubusercontent.com/95455900/192259829-c9f6464a-3f07-46ea-9fa9-2e7381025387.png)

I followed your advice and now the training is more stable and the generated ECGs are much more realistic, although in some samples the beats are more irregular. However, it seems pretty accurate concerning the similarity between the generated samples and my dataset.  


> Conditional generation is ready, I haven't tested it much though. Check out the `AudioDiffusionConditional` in the readme. You have to provide an embedding, ideally not directly text but some learned embedding from a transformer output.

The sample I attached above is text-conditioned, and I used the text "sinus rhythm, normal ecg" to generate the sample. However, after training, it seems to me that the generated samples aren't able to follow the text conditions very well. For example, generating a sample with "bradycardia" (slow bpm) or "tachycardia" (fast bpm) generates similar samples with the above. 

You mentioned in the AudioDiffusionConditional that I should be using a embedding and not directly text. I am currently using text tokenized with BPE and setting embedding_features to 1 for text-conditioning. Will this not work properly? The reason for doing this is because I used tokenized text for other baseline models, and wanted to make a fair comparison.

**Reply by flavioschneider** [reply]
Very exciting progress! The way you're doing text conditioning is very unlikely to work. Embeddings should be vectors that can be attended to internally with cross attention. Ideally, you would train or use a pretrained transformer to interpret the grammatical content of the text, but in your case it seems that it's enough to correlate a generated ecg with the input text (e.g. if there's the word "tachycardia", the model wouldn't have to understand the grammatical structure of the sentence, but just correlate that word with a given ecg). 

To do so, you can add an embedding block before the `AudioDiffusionConditional` model, something like: 

```py
# How many words (numbers) are in your BPE vocabulary 
num_possible_words = 100
# The size of the vector, I'd suggest 768 it's quite standard
embedding_features = 768
# Maximum number of word vectors per waveform
max_length = 6 
# Build trainable embedding 
embedder = nn.Embedding(num_possible_words, embedding_features)

model = AudioDiffusionConditional(
    in_channels=12,
    embedding_max_length=max_length, 
    embedding_features=embedding_features,
    embedding_mask_proba=0.1 # Conditional dropout of batch elements
    patch_factor=4,
    multipliers=[1, 2, 4, 4, 4],
    factors=[4, 2, 2, 2],
    num_blocks=[2, 2, 2, 2],
    attentions=[0, 1, 1, 1, 1],
)

# Then during training 

# E.g. assume that the sentence "sinus rhythm, normal ecg" is tokenized to [1, 5, 8, 3, 0, 0] (in this example max length is 6, change it to whatever you want) 
tokens = torch.tensor([[1, 5, 8, 3, 0, 0]])
# Convert to learned vectors of shape (1, 6, 768) = (batch_size, max_length, embedding_features) 
embedding = embedder(tokens) 
ecgs = torch.randn(1, 12, 8192)
model(ecgs, embedding=embedding)
```
